tests/integration/upload.tap.js

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2021 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const tap = require('tap')
+const { execSync } = require('child_process')
+const upload = require('../../lib/upload')
+const nock = require('nock')
+
+tap.test('upload', function(t) {
+  nock('https://nr-downloads-main.s3.amazonaws.com:443')
+    .put(/nodejs_agent/)
+    .reply(200)
+  execSync('rm -rf ./build/Release/*')
+  execSync(`node ./lib/pre-build build native_metrics`)
+  upload('native_metrics', t.end)
+})
+
+tap.test('upload failure', function(t) {
+  nock('https://nr-downloads-main.s3.amazonaws.com:443')
+    .put(/nodejs_agent/)
+    .reply(404)
+  execSync('rm -rf ./build/Release/*')
+  execSync(`node ./lib/pre-build build native_metrics`)
+  upload('native_metrics', (err) => {
+    t.ok(err, 'should error when not returning 200')
+    t.end()
+  })
+})
```

Comment:
I'd probably disable external calls prior to each test and then re-enable. And maybe have a check for non-called nock endpoints to log and then clean them up (less important for these tests). Risk is minimal for doing anything real but would prevent that and make more obvious anything else going awry.

```
nock.disableNetConnect()
nock.enableNetConnect()
```

I can't remember with nock if we have to explicitly call isDone() or if it will just yell at us for not hitting and endpoint we've setup. I'm assuming you've made these fail so they are fine but mentioning for consideration. Although, I guess the callback responses *should* mostly handle that by getting what you expect/don't expect.

Reply:
good call on disabling to make sure. I verified these tests passed by turning off wifi locally but never know going forward